### PR TITLE
Feature/chai slice

### PR DESCRIPTION
**Pull request by rfhaque**
Implementation of slices

### CODE REVIEW

**Comment by davidbeckingsale** [comment]
@rfhaque - it doesn't look like this is up to date with the latest changes in `develop`. Can you please check that?

**Comment by rfhaque** [comment]
@davidbeckingsale : seems up to date with develop. Which file do you find suspect?

**Comment by davidbeckingsale** [comment]
Whoops! I was mixing up the _thin file with the real one. This looks good.

@robinson96 - can you take a look?

File: src/chai/ManagedArray.inl [hunk 1]
```diff
@@ -133,47 +149,74 @@ template<typename T>
...
+    CHAI_LOG("ManagedArray", "m_active_ptr reallocated at address: " << m_active_pointer);
+  }
 }
 
 template<typename T>
 CHAI_INLINE
 CHAI_HOST void ManagedArray<T>::free()
 {
-  m_resource_manager->free(m_pointer_record);
+  if(!m_is_slice) m_resource_manager->free(m_pointer_record);
```
Comment (robinson96) [inline src/chai/ManagedArray.inl]: It's probably worth doing a LOG message if someone attempts to free a slice - it's good that it's not  free'd here, but it's probably indicative of a bug in the main program.

File: tests/managed_array_tests.cpp [hunk 2]
```diff
@@ -369,6 +412,28 @@ CUDA_TEST(ManagedArray, IncrementDecrementFromHostOnDeviceUM) {
...
+CUDA_TEST(ManagedArray, PickandSetSliceDeviceToDeviceUM) {
+  chai::ManagedArray<int> array(10, chai::UM);
+  chai::ManagedArray<int> sl1 = array.slice(0,5);
+  chai::ManagedArray<int> sl2 = array.slice(5,5);
+
+  forall(cuda(), 0, 10, [=] __device__ (int i) {
+      array[i] = i;
+  });
+
+  forall(cuda(), 0, 5, [=] __device__ (int i) {
```
Comment (robinson96) [inline tests/managed_array_tests.cpp]: Is there a way to write the unit test to check that only one device to host transfer occurs before the sequential loop in this test?

**Pull request merged**